File: bipartite_graphs/cpp/bipartite_graphs.cpp
#include <fstream>
#include <iostream>
#include <queue>
#include <sstream>
#include <vector>

bool 
IsBipartite(const std::vector<std::vector<int>>& graph, int start)
{
    size_t n = graph.size();
    std::vector<int> colour(n, -1); // -1: uncoloured, 0: colour A, 1: colour B
    std::queue<int> q;
    q.push(start);
    colour[start] = 0;

    while (!q.empty())
    {
        int node = q.front();
        q.pop();

        for (int neighbour : graph[node])
        {
            if (colour[neighbour] == -1)
            {
                colour[neighbour] = 1 - colour[node]; // Assigns the opposite colour
                q.push(neighbour);
            }
            else if (colour[neighbour] == colour[node])
            {
                return false;
            }
        }
    }

    return true;
}

int main()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);

    std::ifstream inputFile("bipartite_graph_input_2.txt");
    if (!inputFile)
    {
        std::cout << "Error: Unable to open input file" << std::endl;
        return 1;
    }

    int nodes, edges;
    inputFile >> nodes >> edges;

    std::vector<std::vector<int>> graph(nodes);
    for (int i = 0; i < edges; ++i)
    {
        int from_node, to_node;
        inputFile >> from_node >> to_node;
        graph[from_node].push_back(to_node);
        graph[to_node].push_back(from_node);
    }

    inputFile.close();

    bool is_bipartite = true;
    std::vector<int> colour(nodes, -1);
    for (int i = 0; i < nodes; ++i)
    {
        if (colour[i] == -1 && !IsBipartite(graph, i))
        {
            is_bipartite = false;
            break;
        }
    }

    if (is_bipartite)
    {
        std::cout << "The graph is bipartite" << std::endl;
    }
    else
    {
        std::cout << "The graph is not bipartite" << std::endl;
    }

    return 0;
}

File: bipartite_graphs/python/bipartite_graphs.py
from collections import deque

def IsBipartite(graph, start):
    n = len(graph)
    colour = [-1] * n  # -1: uncoloured, 0: colour A, 1: colour B
    q = deque([start])
    colour[start] = 0

    while q:
        node = q.popleft()
        for neighbour in graph[node]:
            if colour[neighbour] == -1:
                colour[neighbour] = 1 - colour[node]  # Assigns the opposite colour
                q.append(neighbour)
            elif colour[neighbour] == colour[node]:
                return False

    return True


if __name__ == "__main__":
    try:
        with open("bipartite_graph_input_3.txt", "r") as input_file:
            nodes, edges = map(int, input_file.readline().split())

            graph = [[] for _ in range(nodes)]
            for _ in range(edges):
                from_node, to_node = map(int, input_file.readline().split())
                graph[from_node].append(to_node)
                graph[to_node].append(from_node)

    except IOError:
        print("Error: Unable to open input file")
        exit()

    is_bipartite = True
    colour = [-1] * nodes
    for i in range(nodes):
        if colour[i] == -1 and not IsBipartite(graph, i):
            is_bipartite = False
            break

    if is_bipartite:
        print("The graph is bipartite")
    else:
        print("The graph is not bipartite")
